Task#6 Maybe

KroosT opened the pull request:
Created class _Maybe_ which may contain or not contain value.
It contain _isDefined()_ method that returns _true_ if value is present and _getOrThrow()_ method that throws exception if _isDefined() == false_ and returns value otherwise.

KroosT committed:
Task #6 Maybe moved from another branch.

KroosT committed:
Modified Maybe class.

Now it has some sense.

KroosT committed:
Changed Maybe to be module.

KroosT committed:
Modified Maybe to be package instead of module.

KroosT committed:
Class modified.

Add empty() method.
Renamed create() method to just().

KroosT committed:
Added Validator class.

KroosT committed:
.gitignore changed

vmatelsky commented on Tasks/src/Maybe/Maybe.java:
```diff
@@ -0,0 +1,40 @@
+        return new Maybe<>((T) null);
+    }
+
+    public T getValueOrThrow() {
```
what about `getOrThrow`?

vmatelsky commented on Tasks/src/Maybe/MaybeTest.java:
```diff
@@ -0,0 +1,32 @@
+    private final Maybe<Integer> mNullMaybe = Maybe.empty();
+
+    @Test
+    public void testGet() throws Exception {
```
split into two tests with meaningfull names

KroosT committed:
Classes modified.

Renamed getValueOrThrow() method to getOrThrow().
Split test method into two. Renamed this two method's names
to be more descriptive.

vmatelsky commented on Tasks/src/Maybe/MaybeTest.java:
```diff
@@ -0,0 +1,40 @@
+
+public class MaybeTest {
+
+    private final Maybe<Integer> mIntegerMaybe = Maybe.just(229);
```
move these inside tests

vmatelsky commented on Tasks/src/Maybe/MaybeTest.java:
```diff
@@ -0,0 +1,40 @@
+    public void testGetOfIntegerMaybe() throws Exception {
+        Integer integerValue = null;
+
+        if (mIntegerMaybe.isDefined()) {
```
`isDefined` should be tested in two separate tests

vmatelsky commented on Tasks/src/Maybe/MaybeTest.java:
```diff
@@ -0,0 +1,40 @@
+        Integer integerValue = null;
+
+        if (mIntegerMaybe.isDefined()) {
+            integerValue = mIntegerMaybe.getOrThrow();
```
same for `getOrThrow`

KroosT committed:
Tests improved.

vmatelsky commented on Tasks/src/Maybe/MaybeTest.java:
```diff
@@ -1,40 +1,35 @@
-        Assert.assertEquals(integerValue, Integer.valueOf(229));
+    public void testIfDefinedReturnsTrue() {
+        final Maybe<Integer> integerMaybe = Maybe.just(229);
+        assertEquals(integerMaybe.isDefined(), true);
```
`assertTrue` ?

vmatelsky commented on Tasks/src/Maybe/MaybeTest.java:
```diff
@@ -1,40 +1,35 @@
-        }
-
-        Assert.assertEquals(integerValue, Integer.valueOf(229));
+    public void testIfDefinedReturnsTrue() {
```
naming

vmatelsky commented on Tasks/src/Maybe/MaybeTest.java:
```diff
@@ -1,40 +1,35 @@
-        }
+    public void testIfDefinedReturnsFalse() {
+        final Maybe<Integer> emptyMaybe = Maybe.empty();
+        assertEquals(emptyMaybe.isDefined(), false);
```
`assertElse`

vmatelsky commented on Tasks/src/Maybe/MaybeTest.java:
```diff
@@ -1,40 +1,35 @@
+    @Test
+    public void testGetOrThrowReturnsValue() throws Exception {
+        final Maybe<Integer> integerMaybe = Maybe.just(229);
+        assertEquals(integerMaybe.getOrThrow(), Integer.valueOf(229));
```
`Integer.valueOf` why do we need it here?

KroosT replied:
Because _getOrThrow()_ returns Integer value.

KroosT committed:
Better naming. Used more readable asserts.

vmatelsky commented on Tasks/src/Maybe/Maybe.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    public static <T> Maybe<T> empty() {
+        return new Maybe<>((T) null);
```
could we avoid creating new instance every time this method called?

KroosT committed:
Avoid creating new instance each time method calls

vmatelsky commented:
LGTM
